proxy/grpcproxy/lease.go

```diff
@@ -47,41 +98,252 @@ func (lp *leaseProxy) LeaseTimeToLive(ctx context.Context, rr *pb.LeaseTimeToLiv
+
+	ctx context.Context
+
+	cancel func()
+	// errCh must receive a non-nil error because that error
+	// will be returned to the calling client to indicate a failure
+	// happened at the backend
+	errCh chan error
+}
+
+func (lps *leaseProxyStream) recvLoop() error {
+	for lps.ctx.Err() == nil {
+		rr, err := lps.stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
 		if err != nil {
 			return err
 		}
+		lps.serveKeepAliveRequest(rr)
+	}
+	return lps.ctx.Err()
+}
+
+func (lps *leaseProxyStream) serveKeepAliveRequest(rr *pb.LeaseKeepAliveRequest) {
+	lps.mu.Lock()
+	defer lps.mu.Unlock()
+	canSendRespC, ok := lps.keepAliveLeases[rr.ID]
+	if !ok {
+		canSendRespC = &atomicCounter{}
+		lps.keepAliveLeases[rr.ID] = canSendRespC
+		lps.wg.Add(1)
+		go func() {
+			defer lps.wg.Done()
+			lps.toErrCh(lps.keepAliveLoop(rr.ID))
+		}()
+	}
+	canSendRespC.add(1)
+}
+
+func (lps *leaseProxyStream) keepAliveLoop(leaseID int64) error {
+	cctx, ccancel := context.WithCancel(lps.ctx)
+	defer func() {
+		ccancel()
+	}()
+	respc, err := lps.leaseCli.KeepAlive(cctx, clientv3.LeaseID(leaseID))
+	if err != nil {
+		return err
 	}
+	var canSendRespC *atomicCounter
+	lps.mu.RLock()
+	canSendRespC = lps.keepAliveLeases[leaseID]
+	lps.mu.RUnlock()
+	// ticker fires and exits keepAlive loop when the keepAlive
+	// loop hasn't receive keepAlive req for a TTL amount of time.
+	var ticker <-chan time.Time
+	for {
+		select {
+		case <-cctx.Done():
+			return cctx.Err()
+		case <-ticker:
+			lps.mu.Lock()
+			// if there are outstanding keepAlive reqs at the moment of ticker firing,
+			// don't close keepAliveLoop(), let it continuing to process the KeepAlive reqs.
+			if canSendRespC.get() > 0 {
+				lps.mu.Unlock()
+				ticker = make(<-chan time.Time)
+				continue
+			}
+			delete(lps.keepAliveLeases, leaseID)
+			lps.mu.Unlock()
+			return nil
+		case rp, ok := <-respc:
+			if !ok {
+				lps.mu.Lock()
+				delete(lps.keepAliveLeases, leaseID)
+				lps.mu.Unlock()
+				if canSendRespC.get() == 0 {
+					return nil
+				}
+				ttlResp, err := lps.leaseCli.TimeToLive(cctx, clientv3.LeaseID(leaseID))
+				if err != nil {
+					return err
+				}
+				r := &pb.LeaseKeepAliveResponse{
+					Header: ttlResp.ResponseHeader,
+					ID:     int64(ttlResp.ID),
+					TTL:    ttlResp.TTL,
+				}
+				lps.respc <- r
+				return nil
+			}
+			if canSendRespC.get() == 0 {
+				continue
+			}
+			ticker = time.After(time.Duration(rp.TTL) * time.Second)
+			r := &pb.LeaseKeepAliveResponse{
+				Header: rp.ResponseHeader,
+				ID:     int64(rp.ID),
+				TTL:    rp.TTL,
+			}
+			lps.respc <- r
+			canSendRespC.add(-1)
+		}
+	}
+}
+
+func (lps *leaseProxyStream) sendLoop() error {
+	for {
+		lrp, ok := <-lps.respc
+		if !ok {
+			return nil
+		}
+		lps.stream.Send(lrp)
+	}
+}
+
+// toErrCh only sends non nil error to errCh to ensure
+// only the non-nil error will be returned to the calling client
+func (lps *leaseProxyStream) toErrCh(err error) {
```

Comment:
the errch is not buffered. what if it receives multiple errors at the same time and blocks goroutines that sends the error. e.g multiple keepAliveLoop goroutines sends error at the same time. My original thought is that the first error wins and will be sent back to client and rest of errors will be ignored.